core/src/Streamly/Internal/FileSystem/Path/Common.hs

```diff
@@ -269,63 +280,148 @@ hasDrive a
+-- someone uses "./x" they explicitly mean x in the current directory whereas
+-- just "x" can be taken to mean a path segment without any specific root.
+-- However, in typed paths the programmer can convey the meaning whether they
+-- mean it as a segment or a rooted path. So even "./x" can potentially be used
+-- as a segment which can just mean "x".
 --
--- On Windows:
--- * @C:\\@ local absolute
--- * @C:@ local relative
--- * @\\@ local relative to current drive root
--- * @\\\\@ UNC network location
+-- XXX For the untyped Path we can allow appending "./x" to other paths. We can
+-- leave this to the programmer. In typed paths we can allow "./x" in segments.
+--
+-- XXX C:\\ is invalid, \\share\ is invalid?
+-- XXX Empty path can be taken to mean "." except in case of UNC paths
+
+-- | Rooted paths on Posix and Windows,
+-- * @/...@ a path starting with a separator
+-- * @.@ current dir
+-- * @./...@ a location relative to current dir
+--
+-- Rooted paths on Windows:
+-- * @C:@ local drive cur dir location
+-- * @C:a\\b@ local drive relative to cur dir location
+-- * @C:\\@ local drive absolute location
+-- * @\\@ local path relative to current drive
+-- * @\\\\share\\@ UNC network location
 -- * @\\\\?\\C:\\@ Long UNC local path
 -- * @\\\\?\\UNC\\@ Long UNC server location
 -- * @\\\\.\\@ DOS local device namespace
 -- * @\\\\??\\@ DOS global namespace
 --
--- * @C:file@ a path relative to curdir.
 isLocation :: (Unbox a, Integral a) => OS -> Array a -> Bool
 isLocation Posix a =
-    -- Assuming path is not empty.
-    isSeparator Posix (wordToChar (Array.getIndexUnsafe 0 a))
-        || isCurDirRelativeLocation a
+    isRelativeCurDrive Posix a
+        || isRelativeCurDir a
 isLocation Windows a =
-    isLocationDrive a
-        || isCurDriveRelativeLocation a
+    isRelativeCurDrive Windows a
+        || isRelativeCurDir a
         || hasDrive a -- curdir-in-drive relative, drive absolute
-        || isAbsoluteUNCLocation a
-        || isCurDirRelativeLocation a
+        || isAbsoluteUNC a
 
 isSegment :: (Unbox a, Integral a) => OS -> Array a -> Bool
 isSegment os = not . isLocation os
 
+------------------------------------------------------------------------------
+-- File or Dir
+------------------------------------------------------------------------------
+
+-- | Returns () if the path can be a valid file, otherwise throws an
+-- exception.
+maybeFile :: (MonadThrow m, Unbox a, Integral a) => OS -> Array a -> m ()
+maybeFile os arr = do
+    s1 <-
+            Stream.toList
+                $ Stream.take 3
+                $ Stream.takeWhile (not . isSeparator os)
+                $ fmap wordToChar
+                $ Array.readRev arr
+    -- XXX On posix we just need to check last 3 bytes of the array
+    -- XXX Display the path in the exception messages.
+    case s1 of
+        [] -> throwM $ InvalidPath "A file name cannot have a trailing separator"
```

Comment:
Even an empty path could result in this. The error message does not indicate that.
So, `maybeFile` would fail if the path ends with
1. `.`
2. `..`

Reply:
Empty path is not a valid path. 

> So, maybeFile would fail if the path ends with

yes.